cli/napi/uv.rs

```diff
@@ -0,0 +1,237 @@
+      let mut attr = attr.assume_init();
+      let attr_ptr = addr_of_mut!(attr);
+      assert_ok(libc::pthread_mutexattr_settype(
+        attr_ptr,
+        libc::PTHREAD_MUTEX_ERRORCHECK,
+      ));
+      let err = libc::pthread_mutex_init(lock, attr_ptr);
+      assert_ok(libc::pthread_mutexattr_destroy(attr_ptr));
+      if libc::EDOM > 0 {
+        err
+      } else {
+        -err
+      }
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_lock(lock: *mut uv_mutex_t) {
+    unsafe {
+      assert_ok(libc::pthread_mutex_lock(lock));
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_unlock(lock: *mut uv_mutex_t) {
+    unsafe {
+      assert_ok(libc::pthread_mutex_unlock(lock));
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_destroy(lock: *mut uv_mutex_t) {
+    unsafe {
+      assert_ok(libc::pthread_mutex_destroy(lock));
+    }
+  }
+}
+#[cfg(windows)]
+mod mutex {
+  use super::*;
+  use windows_sys::Win32::System::Threading as win;
+  type uv_mutex_t = win::CRITICAL_SECTION;
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_init(lock: *mut uv_mutex_t) -> c_int {
+    unsafe {
+      win::InitializeCriticalSection(lock);
+    }
+    0
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_lock(lock: *mut uv_mutex_t) {
+    unsafe {
+      win::EnterCriticalSection(lock);
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_unlock(lock: *mut uv_mutex_t) {
+    unsafe {
+      win::LeaveCriticalSection(lock);
+    }
+  }
+
+  #[no_mangle]
+  unsafe extern "C" fn uv_mutex_destroy(lock: *mut uv_mutex_t) {
+    unsafe {
+      win::DeleteCriticalSection(lock);
+    }
+  }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+#[allow(dead_code)]
+enum uv_handle_type {
+  UV_UNKNOWN_HANDLE = 0,
+  UV_ASYNC,
+  UV_CHECK,
+  UV_FS_EVENT,
+  UV_FS_POLL,
+  UV_HANDLE,
+  UV_IDLE,
+  UV_NAMED_PIPE,
+  UV_POLL,
+  UV_PREPARE,
+  UV_PROCESS,
+  UV_STREAM,
+  UV_TCP,
+  UV_TIMER,
+  UV_TTY,
+  UV_UDP,
+  UV_SIGNAL,
+  UV_FILE,
+  UV_HANDLE_TYPE_MAX,
+}
+
+const UV_HANDLE_SIZE: usize = 96;
+
+#[repr(C)]
+struct uv_handle_t {
+  // public members
+  pub data: *mut c_void,
+  pub r#loop: *mut uv_loop_t,
+  pub r#type: uv_handle_type,
+
+  _padding: [MaybeUninit<u8>; const {
+    UV_HANDLE_SIZE
+      - size_of::<*mut c_void>()
+      - size_of::<*mut uv_loop_t>()
+      - size_of::<uv_handle_type>()
+  }],
+}
+
+#[cfg(unix)]
+const UV_ASYNC_SIZE: usize = 128;
+
+#[cfg(windows)]
+const UV_ASYNC_SIZE: usize = 224;
+
+#[repr(C)]
+struct uv_async_t {
+  // public members
+  pub data: *mut c_void,
+  pub r#loop: *mut uv_loop_t,
+  pub r#type: uv_handle_type,
+  // private
+  async_cb: uv_async_cb,
+  work: napi_async_work,
+  _padding: [MaybeUninit<u8>; const {
+    UV_ASYNC_SIZE
+      - size_of::<*mut c_void>()
+      - size_of::<*mut uv_loop_t>()
+      - size_of::<uv_handle_type>()
+      - size_of::<uv_async_cb>()
+      - size_of::<napi_async_work>()
+  }],
+}
```

Comment:
maybe add smth like this just to be sure.
```rs
const _: () = assert!(std::mem::size_of::<uv_async_t>() == UV_ASYNC_SIZE)
```
or maybe a test which compares it to the `uv_sys_lite` type?

Reply:
Done